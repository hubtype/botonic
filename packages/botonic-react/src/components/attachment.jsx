import React from 'react'
import AttachmentIcon from '../assets/attachment-icon.svg'
import { staticAsset } from '../utils'
export const Attachment = ({ onChange, accept }) => (
  <div
    style={{
      display: 'flex',
      alignItems: 'center',
      paddingRight: 15,
    }}
  >
    <label htmlFor='attachment'>
      <img
        src={staticAsset(AttachmentIcon)}
        style={{ cursor: 'pointer', marginTop: 4 }}
      />
    </label>
    <input
      type='file'
      name='file'
      id='attachment'
      style={{ display: 'none' }}
      onChange={onChange}
      accept={accept}
    ></input>
  </div>
)
