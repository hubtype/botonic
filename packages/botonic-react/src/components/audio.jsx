import { INPUT, isBrowser } from '@botonic/core'
import React from 'react'

import { ROLES } from '../constants'
import { Message } from './message'

const serialize = audioProps => {
  return { audio: audioProps.src }
}

export const Audio = props => {
  let content = props.children
  if (isBrowser())
    content = (
      <audio style={{ maxWidth: '100%' }} id='myAudio' controls>
        <source src={props.src} type='audio/mpeg' />
        Your browser does not support this audio format.
      </audio>
    )
  return (
    <Message
      role={ROLES.AUDIO_MESSAGE}
      json={serialize(props)}
      {...props}
      type={INPUT.AUDIO}
    >
      {content}
    </Message>
  )
}

Audio.serialize = serialize
