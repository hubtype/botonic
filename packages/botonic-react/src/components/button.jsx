import { INPUT, params2queryString } from '@botonic/core'
import React, { useContext, useState } from 'react'
import styled from 'styled-components'

import { COLORS, WEBCHAT } from '../constants'
import { WebchatContext } from '../contexts'
import { renderComponent } from '../util/react'
import { ButtonsDisabler } from './buttons-disabler'

const StyledButton = styled.button`
  display: flex;
  width: 100%;
  max-height: 80px;
  font-size: 14px;
  text-align: center;
  align-content: center;
  justify-content: center;
  padding: 12px 32px;
  font-family: inherit;
  border: none;
  border: 1px solid ${COLORS.SEASHELL_WHITE};
  cursor: pointer;
  outline: 0;
  border-top-right-radius: ${props => props.top || '0px'};
  border-top-left-radius: ${props => props.top || '0px'};
  border-bottom-right-radius: ${props => props.bottom || '0px'};
  border-bottom-left-radius: ${props => props.bottom || '0px'};
  overflow: hidden;
`

export const Button = props => {
  const {
    webchatState,
    openWebview,
    sendPayload,
    sendInput,
    getThemeProperty,
    updateMessage,
  } = useContext(WebchatContext)
  const [hover, setHover] = useState(false)
  const { theme } = webchatState
  const { autoDisable, disabledStyle } = ButtonsDisabler.resolveDisabling(
    webchatState.theme,
    props
  )
  const handleClick = event => {
    event.preventDefault()
    const type = getThemeProperty(
      WEBCHAT.CUSTOM_PROPERTIES.buttonMessageType,
      INPUT.TEXT
    )
    if (props.webview) openWebview(props.webview, props.params)
    else if (props.path) {
      type == INPUT.POSTBACK
        ? sendPayload(`__PATH_PAYLOAD__${props.path}`)
        : sendInput({
            type: INPUT.TEXT,
            data: String(props.children),
            payload: `__PATH_PAYLOAD__${props.path}`,
          })
    } else if (props.payload) {
      type == INPUT.POSTBACK
        ? sendPayload(props.payload)
        : sendInput({
            type: INPUT.TEXT,
            data: String(props.children),
            payload: props.payload,
          })
    } else if (props.url) {
      window.open(props.url, props.target || '_blank')
    }
    if (props.onClick) props.onClick()
    if (props.setDisabled) {
      props.setDisabled(true)
      const messageToUpdate = webchatState.messagesJSON.filter(
        m => m.id == props.parentId
      )[0]
      const updatedMsg = ButtonsDisabler.getUpdatedMessage(messageToUpdate, {
        autoDisable,
        disabledStyle,
      })
      updateMessage(updatedMsg)
    }
  }

  const renderBrowser = () => {
    const buttonStyle = getThemeProperty(WEBCHAT.CUSTOM_PROPERTIES.buttonStyle)
    const CustomButton = getThemeProperty(
      WEBCHAT.CUSTOM_PROPERTIES.customButton
    )
    if (CustomButton) {
      return (
        <div onClick={e => handleClick(e)}>
          <CustomButton>{props.children}</CustomButton>
        </div>
      )
    }

    const buttonBgColor = hover
      ? getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonHoverBackground,
          COLORS.CONCRETE_WHITE
        )
      : getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonStyleBackground,
          COLORS.SOLID_WHITE
        )
    const buttonTextColor = hover
      ? getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonHoverTextColor,
          COLORS.SOLID_BLACK
        )
      : getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonStyleColor,
          COLORS.SOLID_BLACK
        )

    return (
      <StyledButton
        theme={theme}
        onMouseEnter={() => setHover(true)}
        onMouseLeave={() => setHover(false)}
        onClick={e => handleClick(e)}
        style={{
          ...buttonStyle,
          color: buttonTextColor,
          backgroundColor: buttonBgColor,
          ...(props.disabled && autoDisable && disabledStyle),
        }}
        bottom={props.bottomRadius}
      >
        {props.children}
      </StyledButton>
    )
  }

  const renderNode = () => {
    const disabledProps = ButtonsDisabler.constructNodeProps(props)
    if (props.webview) {
      const Webview = props.webview
      let params = ''
      if (props.params) params = params2queryString(props.params)
      return (
        <button url={`/webviews/${Webview.name}?${params}`} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.path) {
      const payload = `__PATH_PAYLOAD__${props.path}`
      return (
        <button payload={payload} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.payload) {
      return (
        <button payload={props.payload} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.url) {
      return (
        <button url={props.url} target={props.target} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.onClick) {
      return null
    }
    throw new Error('Button missing payload, path, webviews, url or onClick')
  }

  return renderComponent({ renderBrowser, renderNode })
}

Button.serialize = buttonProps => {
  let payload = buttonProps.payload
  if (buttonProps.path) payload = `__PATH_PAYLOAD__${buttonProps.path}`
  return {
    button: {
      payload,
      url: buttonProps.url,
      target: buttonProps.target,
      webview: buttonProps.webview && String(buttonProps.webview),
      title: buttonProps.children && String(buttonProps.children),
      ...ButtonsDisabler.withDisabledProps(buttonProps),
    },
  }
}
