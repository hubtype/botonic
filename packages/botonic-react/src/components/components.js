/**
 * Get the name of an element (instantiated component)
 * @param {React.Element} component
 * @returns {string}
 */
export function getElementName(component) {
  return component.type.name
}

/**
 * Get the name of a component TYPE
 * @param {React.ComponentType} component
 * @returns {string}
 */
export function getComponentTypeName(component) {
  return component.displayName || component.name || 'Component'
}
