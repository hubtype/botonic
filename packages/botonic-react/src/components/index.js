export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export { Location } from './location'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export { WhatsappTemplate } from './whatsapp-template'
