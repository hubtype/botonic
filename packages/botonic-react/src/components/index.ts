export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export { WhatsappTemplate } from './whatsapp-template'
