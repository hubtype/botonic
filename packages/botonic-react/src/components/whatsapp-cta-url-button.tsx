import { INPUT } from '@botonic/core'
import React from 'react'

import { truncateText } from '../util'
import { renderComponent } from '../util/react'
import { generateWebviewUrlWithParams } from '../util/webviews'
import { Message } from './message'
import {
  WHATSAPP_MAX_BODY_CHARS,
  WHATSAPP_MAX_BUTTON_CHARS,
  WHATSAPP_MAX_FOOTER_CHARS,
  WHATSAPP_MAX_HEADER_CHARS,
} from './multichannel/multichannel-utils'
import { whatsappMarkdown } from './multichannel/whatsapp/markdown'

export interface WhatsappCTAUrlButtonCommonProps {
  header?: string
  body: string
  footer?: string
  displayText: string
}

export interface WhatsappCTAUrlButtonUrlProps
  extends WhatsappCTAUrlButtonCommonProps {
  url: string
}

export interface WhatsappCTAUrlButtonWebviewProps
  extends WhatsappCTAUrlButtonCommonProps {
  webview: any
  params?: any
}

export type WhatsappCTAUrlButtonProps =
  | WhatsappCTAUrlButtonUrlProps
  | WhatsappCTAUrlButtonWebviewProps

const serialize = _whatsappCTAUrlButtonProps => {
  // TODO: Implement to have data persistance in localStorage, not needed for this WhatsApp development
  return {}
}

export const WhatsappCTAUrlButton = (props: WhatsappCTAUrlButtonProps) => {
  const renderBrowser = () => {
    // Return a dummy message for browser
    const message = `${JSON.stringify(props)}`
    return (
      <Message
        json={serialize(message)}
        {...props}
        type={INPUT.WHATSAPP_CTA_URL_BUTTON}
      >
        {message}
      </Message>
    )
  }

  const renderNode = () => {
    const validatedProps = {
      ...props,
      header: props.header
        ? truncateText(props.header, WHATSAPP_MAX_HEADER_CHARS)
        : undefined,
      body: truncateText(whatsappMarkdown(props.body), WHATSAPP_MAX_BODY_CHARS),
      footer: props.footer
        ? truncateText(props.footer, WHATSAPP_MAX_FOOTER_CHARS)
        : undefined,
      displayText: truncateText(props.displayText, WHATSAPP_MAX_BUTTON_CHARS),
      url:
        'webview' in props
          ? generateWebviewUrlWithParams(props.webview, props.params)
          : props.url,
    }
    return (
      // @ts-ignore Property 'message' does not exist on type 'JSX.IntrinsicElements'.
      <message {...validatedProps} type={INPUT.WHATSAPP_CTA_URL_BUTTON} />
    )
  }

  return renderComponent({ renderBrowser, renderNode })
}
