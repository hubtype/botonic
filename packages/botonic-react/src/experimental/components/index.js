export { Button } from '../../components/button'
export { Element } from '../../components/element'
export { MessageTemplate } from '../../components/message-template'
export * from '../../components/multichannel'
export { Pic } from '../../components/pic'
export { Reply } from '../../components/reply'
export { ShareButton } from '../../components/share-button'
export { Subtitle } from '../../components/subtitle'
export { Title } from '../../components/title'
export { WebchatSettings } from '../../components/webchat-settings'

// Experimental
export { Audio } from './audio'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Image } from './image'
export { Location } from './location'
export { Text } from './text'
export { Video } from './video'
export { WhatsappTemplate } from './whatsapp-template'
