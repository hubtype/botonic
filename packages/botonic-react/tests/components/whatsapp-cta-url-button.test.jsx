import { expect, test } from '@jest/globals'
import React from 'react'
import TestRenderer from 'react-test-renderer'

import { WhatsappCTAUrlButton } from '../../src/components/whatsapp-cta-url-button'

const renderToJSON = sut => TestRenderer.create(sut).toJSON()

test('renders WhatsappCTAUrlButton component (passing a URL)', () => {
  const props = {
    header: 'This is the header',
    body: 'This is the body',
    footer: 'This is the footer',
    displayText: 'Go to Hubtype',
    url: 'https://www.hubtype.com',
  }

  const tree = renderToJSON(<WhatsappCTAUrlButton {...props} />)
  expect(tree).toMatchSnapshot()
})

test('renders WhatsappCTAUrlButton component (passing a Webview)', () => {
  class MyWebview {}
  const props = {
    header: 'This is the header',
    body: 'This is the body',
    footer: 'This is the footer',
    displayText: 'Go to webview',
    webview: MyWebview,
    params: {
      numbers: '123',
      redirectUri: 'www.some-site.com',
    },
  }

  const tree = renderToJSON(<WhatsappCTAUrlButton {...props} />)
  expect(tree).toMatchSnapshot()
})
