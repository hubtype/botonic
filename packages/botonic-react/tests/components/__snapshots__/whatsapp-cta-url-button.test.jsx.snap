// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders WhatsappCTAUrlButton component (passing a URL) 1`] = `
<message
  body="This is the body"
  displayText="Go to Hubtype"
  footer="This is the footer"
  header="This is the header"
  type="whatsapp-cta-url-button"
  url="https://www.hubtype.com"
/>
`;

exports[`renders WhatsappCTAUrlButton component (passing a Webview) 1`] = `
<message
  body="This is the body"
  displayText="Go to webview"
  footer="This is the footer"
  header="This is the header"
  params={
    {
      "numbers": "123",
      "redirectUri": "www.some-site.com",
    }
  }
  type="whatsapp-cta-url-button"
  url="/webviews/MyWebview?numbers=123&redirectUri=www.some-site.com"
  webview={[Function]}
/>
`;
